Automate BBH pipeline

nilsvu opened the pull request:
## Proposed changes

Run the inspiral automatically after ID generation. Also continue with ringdown automatically. Implements parts of #5021. Things like eccentricity control etc can be added here as well.

Try it like this (e.g. on Wheeler):

```
./bin/spectre bbh generate-id -q 1 -D 16 -w 0.0144 -d ./BbhTest --evolve
```

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Allow 'schedule' to fill parameters in input file metadata

nilsvu committed:
Allow input files to specify a 'Next' Python function to run

nilsvu committed:
Automate BBH pipeline

nilsvu committed:
Run inspiral on 4 nodes by default

Because 1 node is not enough. This can be made smarter later.

knelli2 commented on support/Python/RunNext.py:
```diff
@@ -0,0 +1,126 @@
+        metadata = next(yaml.safe_load_all(open_input_file))
+    if "Next" not in metadata:
+        logger.info(
+            f"The input file lists no 'Next' entrypoint. Nothing to do."
```
`The input file metadata lists no 'Next'...` just to be specific

knelli2 commented on support/Python/__main__.py:
```diff
@@ -35,6 +35,7 @@ def list_commands(self, ctx):
             "render-1d",
             "render-3d",
             "resubmit",
+            "run-next",
             "schedule",
```
I'm worried that people will misinterpret this as running the next segment of an evolution, and not the next portion of a pipeline. It's not obvious to me that you'd have to use `schedule` to run the next segment. No real suggestion at the moment, just wanted to note there might be confusion on this.

nilsvu replied:
Hm all I can think of is to look at the help string:

```
$ ./bin/spectre --help
...
  resubmit               Create the next segment in the SEGMENTS_DIR and...
  run-next               Run the next entrypoint specified in the input...
...
$ ./bin/spectre resubmit -h
Usage: spectre resubmit [OPTIONS] [SEGMENTS_DIRS]...

  Create the next segment in the SEGMENTS_DIR and schedule it
...
$ ./bin/spectre run-next -h
Usage: spectre run-next [OPTIONS] INPUT_FILE_PATH

  Run the next entrypoint specified in the input file metadata
...
```

That makes it pretty clear I think.

knelli2 commented on tests/support/Python/Test_RunNext.py:
```diff
@@ -0,0 +1,64 @@
+            [str(self.input_file_path)],
+            catch_exceptions=False,
+        )
+        self.assertEqual(result.exit_code, 0)
+        self.assertIn(spectre.__version__, result.stdout)
```
Will this test pass in `PY_DEV_MODE`?

nilsvu replied:
yep

knelli2 commented on support/Pipelines/Bbh/InitialData.py:
```diff
@@ -76,6 +80,28 @@ def generate_id(
+            " evolution, etc will be created in the 'pipeline_dir'"
+            " automatically."
+        )
+    if pipeline_dir and not run_dir:
+        run_dir = pipeline_dir / "001_ID"
```
One index or zero index?

Also, not for right now, but once we have eccentricity reduction, we're going to maybe have to change the numbering since we'll need to have multiple ID/Ev rounds

nilsvu replied:
Yep this is just a preliminary hardcoded numbering from 1 to 3. With ecc control etc we can make this dynamic and increment the number with each ecc step.

knelli2 commented on support/Pipelines/Bbh/Ringdown.py:
```diff
@@ -144,6 +158,15 @@ def start_ringdown(
+    help="Directory where steps in the pipeline are created.",
+)
 @scheduler_options
 def start_ringdown_command(**kwargs):
     _rich_traceback_guard = True  # Hide traceback until here
```
[Random line] Is it easy to have some tests that all the pipeline directory stuff works, even if it's just `--no-submit` and checking some input files?

nilsvu replied:
I couldn't come up with a good test of the full pipeline because the steps actually need data from the previous steps. Of course we can and should add integration/regression tests that run the pipeline on a cluster. I added some tests of the generated input files though as you suggested.

knelli2 commented on support/Pipelines/Bbh/Inspiral.py:
```diff
@@ -114,6 +114,16 @@ def start_inspiral(
+        and scheduler_kwargs.get("num_procs") is None
+        and scheduler_kwargs.get("num_nodes") is None
+    ):
+        # Just run on 4 nodes for now, because 1 is surely not enough. We can
+        # make this smarter later (e.g. scale with the number of elements).
+        scheduler_kwargs["num_nodes"] = 4
```
Can you add this to the help string above so users are aware of it?

Also, I think this is something that should be machine specific? 4 wheeler nodes are not equivalent to 4 Caltech HPC nodes. Not saying to implement this now, 4 seems like an ok default for all our clusters, but for future consideration.

nilsvu replied:
Added to the docs. Yes this has to scale with the node size somehow. Maybe we can specify the number of cores (rather than nodes) and make the scheduler select the number of nodes.

knelli2 reviewed (changes requested):
Overall looks good! I think this is coming together nicely!

nilsvu committed:
Handle path to default submit script better

Also use the default submit script when the 'schedule'
function is invoked in Python.

knelli2 reviewed (commented):
Ok looks good. Sqash
